File: src/omnia/data/structure/rx/ObservableDirectedGraph.java
package omnia.data.structure.rx;

import io.reactivex.Flowable;
import omnia.data.structure.DirectedGraph;
import omnia.data.structure.Set;

public interface ObservableDirectedGraph<E> extends DirectedGraph<E>, ObservableGraph<E> {
  
  @Override
  ObservableChannels<E> observe();

  interface ObservableChannels<E> extends ObservableGraph.ObservableChannels<E> {

    @Override
    Flowable<? extends DirectedGraph<E>> states();

    @Override
    Flowable<? extends MutationEvent<E>> mutations();
  }

  interface MutationEvent<E> extends ObservableGraph.MutationEvent<E> {

    @Override
    DirectedGraph<E> state();

    @Override
    Set<? extends ObservableGraph.GraphOperation<E>> operations();
  }
}

File: omnia/src/commonMain/omnia/data/structure/observable/ObservableDirectedGraph.kt
package omnia.data.structure.observable

import com.badoo.reaktive.observable.Observable
import omnia.data.structure.DirectedGraph
import omnia.data.structure.Set
import omnia.data.structure.observable.ObservableGraph.GraphOperation

interface ObservableDirectedGraph<E : Any> : DirectedGraph<E>, ObservableGraph<E> {

  override val observables: Observables<E>

  interface Observables<E : Any> : ObservableGraph.Observables<E> {

    override val states: Observable<DirectedGraph<E>>

    override val mutations: Observable<MutationEvent<E>>
  }

  interface MutationEvent<E : Any> : ObservableGraph.MutationEvent<E> {

    override val state: DirectedGraph<E>

    override val operations: Set<out GraphOperation<E>>
  }
}